[a11y] Semantics flag refactor step 2: embedder part

hannah-hyj opened the pull request:
issue: https://github.com/flutter/flutter/issues/166101, overall goal is to update semantics flag to be a struct/class to support more than 32 flags.  


step 1:  https://github.com/flutter/flutter/pull/167421 updated semantics_node.h and  dart:ui
step 2(this PR): Update Embedder part to use a struct instead of a int bit mask.
step 3: https://github.com/flutter/flutter/pull/167771 Update Framework use the SemanticsFlags class instead of bitmask
step 4 https://github.com/flutter/flutter/pull/168852 Update web engine to use the new class and update SemanticsUpdateBuilder.updateNode to pass the new class instead of bitmask

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

hannah-hyj committed:
Update embedder_semantics_update.cc

Update embedder_semantics_update.cc

embedder

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -1594,6 +1628,9 @@ typedef struct {
   const FlutterStringAttribute** decreased_value_attributes;
+  // The set of semantics flags associated with this node. Prefer to use this
+  // over `flags`.
+  FlutterSemanticsFlags* flags2;
```
Could we update flag's comment with something that indicates embedders should move away from it?

cbracken replied:
If the goal is to move off `flags` and replace it with `flags2`, then:
1. Definitely add a comment telling the reader to migrate to `flags2`
2. Consider renaming `flags` to `flags__unused__` to force anyone building against the embedder to migrate.

Even with the field rename, We _must_ continue populating the existing `flags` field from the new struct so that we don't break existing embedders who drop in a new engine shared lib.

loic-sharma replied:
@hannah-hyj FYI, this still needs to be done :)

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -258,6 +258,40 @@ typedef enum {
   kFlutterSemanticsFlagIsRequired = 1 << 30,
 } FlutterSemanticsFlag;
 
+typedef struct {
```
This will need a `struct_size` member. See: https://github.com/flutter/flutter/blob/b8aa0fd8468bc419d53b95345721d28e332a8212/engine/src/flutter/shell/platform/embedder/embedder.h#L12-L50

hannah-hyj replied:
Does that mean we can never change the size of the struct even if we have more flags later?

@chunhtai  If you have to take a guess, how many semantics flags will we eventually need?

loic-sharma replied:
No you can add more members in the future, but embedders will use `struct_size` to check whether it can access that new member.

hannah-hyj replied:
still, if we can't change the size of the struct, we need to predict how many semantics flags we need in the future.

cbracken replied:
`struct_size` is always populated with `sizeof(FlutterSemanticsFlags)`, so it will increase as fields are added. 

This is a bit related to why fields must always be added at the end of a struct and why we can never reorder fields. Embedders built against older versions of the engine will be reading these fields from the memory offsets within the struct from the engine version they built against. If the embedder author drops in a new engine shared lib, those offsets must still match correctly.

hannah-hyj replied:
thanks for explaining this!

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -258,6 +258,40 @@ typedef enum {
 } FlutterSemanticsFlag;
 
+typedef struct {
+  bool hasCheckedState;
```
Could we snake case these? Something like:

```suggestion
  bool has_checked_state;
```

cbracken replied:
Yes -- these must use C/C++ style. So is_checked, has_checked_state, etc.

chunhtai commented on engine/src/flutter/shell/platform/common/accessibility_bridge_unittests.cc:
```diff
@@ -241,9 +243,11 @@ TEST(AccessibilityBridgeTest, DoesNotAssignEditableRootToSelectableText) {
-  root.flags = static_cast<FlutterSemanticsFlag>(
-      FlutterSemanticsFlag::kFlutterSemanticsFlagIsTextField |
-      FlutterSemanticsFlag::kFlutterSemanticsFlagIsReadOnly);
+  FlutterSemanticsFlags flags = FlutterSemanticsFlags{0};
```
do we need to call the constructor? I think just
```
FlutterSemanticsFlags flags;
```
will be fine?

chunhtai replied:
also why 0?

hannah-hyj replied:
I'm calling the constructor here to make sure the flags are all initialized to false  by default.

cbracken replied:
A little more idiomatic:
```cc
FlutterSemanticsFlags flags = FlutterSemanticsFlags{};
```

And yes; as @hannah-hyj says, this is important since locals are uninitialised in C++, so without this, it'll be populated with whatever memory happens to be on the stack at the time.

chunhtai reviewed (commented):
LGTM if also LGT @loic-sharma

cbracken commented on engine/src/flutter/shell/platform/embedder/embedder_semantics_update.cc:
```diff
@@ -279,6 +320,7 @@ void EmbedderSemanticsUpdate2::AddNode(const SemanticsNode& node) {
       increased_value_attributes.attributes,
       decreased_value_attributes.count,
       decreased_value_attributes.attributes,
+      &(flags_.back()),
```
No need for the extra brackets, just:
```suggestion
      &flags_.back(),
```

cbracken commented on engine/src/flutter/shell/platform/embedder/embedder_semantics_update.h:
```diff
@@ -66,6 +66,7 @@ class EmbedderSemanticsUpdate2 {
   std::vector<FlutterSemanticsNode2> nodes_;
   std::vector<FlutterSemanticsNode2*> node_pointers_;
   std::vector<FlutterSemanticsCustomAction2> actions_;
+  std::vector<FlutterSemanticsFlags> flags_;
```
nit: This is safe as-is, but I'd put this at the end to avoid splitting up `actions_` and `action_pointers_`.

cbracken commented on engine/src/flutter/shell/platform/embedder/embedder_semantics_update.cc:
```diff
@@ -126,6 +126,44 @@ FlutterSemanticsFlag SemanticsFlagsToInt(const SemanticsFlags& flags) {
+  dest.hasRequiredState = source.hasRequiredState;
+  dest.isRequired = source.isRequired;
+
+  return dest;
```
Instead, use designated initialiser syntax:
```cc
const FlutterSemanticsFlags convertToFlutterSemanticsFlags(
    const SemanticsFlags& source) {
  return {
    .has_checked_state = source.hasCheckedState,
    .is_checked = source.isChecked,
    // ...
    .is_required = source.isRequired,
  };
}
```

I imagine in both cases the compiler is clever enough to apply return value optimisation, which ensures the return object is constructed directly in the memory address of the caller rather than copied, but this way is more idiomatic and avoids declaring a local at all.

cbracken commented on engine/src/flutter/shell/platform/embedder/embedder_semantics_update.cc:
```diff
@@ -126,6 +126,44 @@ FlutterSemanticsFlag SemanticsFlagsToInt(const SemanticsFlags& flags) {
   return static_cast<FlutterSemanticsFlag>(result);
 }
 
+const FlutterSemanticsFlags convertToFlutterSemanticsFlags(
```
In C++ style, functions and methods are `UpperCamelCase`, except (optionally) for simple getters, which are `lower_snake_case`.

cbracken replied:
It's fine to return const pointers and references, but in C++ it's unconventional to return value objects as `const`. If the caller wants to prevent modifications to the returned object it should mark the variable as const at the call site:
```cc
const FlutterSemanticsFlags flags = ConvertToFlutterSemanticsFlags(source);
```

cbracken replied:
Since this function isn't used outside this translation unit (file), it should be `static` -- or, in more modern C++ in an anonymous namespace, usually at the top of the file. So:
```cc
namespace {
FlutterSemanticsFlags FlutterSemanticsFlagsFromSemanticsFlags(const SemanticsFlags& source) {
  // ...
}
}  // namespace

cbracken commented on engine/src/flutter/shell/platform/common/accessibility_bridge.cc:
```diff
@@ -435,7 +430,7 @@ void AccessibilityBridge::SetBooleanAttributesFromFlutterUpdate(
     const SemanticsNode& node) {
   FlutterSemanticsAction actions = node.actions;
-  FlutterSemanticsFlag flags = node.flags;
+  FlutterSemanticsFlags flags = *(node.flags);
```
Ditto here.

cbracken commented on engine/src/flutter/shell/platform/common/accessibility_bridge.cc:
```diff
@@ -462,15 +454,13 @@ void AccessibilityBridge::SetBooleanAttributesFromFlutterUpdate(
     ui::AXNodeData& node_data,
     const SemanticsNode& node) {
-  FlutterSemanticsFlag flags = node.flags;
+  FlutterSemanticsFlags flags = *(node.flags);
```
And here.

cbracken commented on engine/src/flutter/shell/platform/common/accessibility_bridge.cc:
```diff
@@ -548,13 +535,12 @@ void AccessibilityBridge::SetTooltipFromFlutterUpdate(
-  //    in the tree.
-  if (flags & FlutterSemanticsFlag::kFlutterSemanticsFlagIsTextField &&
-      flags & FlutterSemanticsFlag::kFlutterSemanticsFlagIsFocused) {
+  FlutterSemanticsFlags flags = *(node.flags);
```
And here.

cbracken commented on engine/src/flutter/shell/platform/common/accessibility_bridge_unittests.cc:
```diff
@@ -209,9 +209,11 @@ TEST(AccessibilityBridgeTest, CanHandleSelectionChangeCorrectly) {
-  root.flags = static_cast<FlutterSemanticsFlag>(
-      FlutterSemanticsFlag::kFlutterSemanticsFlagIsTextField |
-      FlutterSemanticsFlag::kFlutterSemanticsFlagIsFocused);
+  FlutterSemanticsFlags flags = FlutterSemanticsFlags{0};
```
Style nit:
```suggestion
  FlutterSemanticsFlags flags = FlutterSemanticsFlags{};
```

cbracken commented on engine/src/flutter/shell/platform/common/accessibility_bridge_unittests.cc:
```diff
@@ -209,9 +209,11 @@ TEST(AccessibilityBridgeTest, CanHandleSelectionChangeCorrectly) {
-  root.flags = static_cast<FlutterSemanticsFlag>(
-      FlutterSemanticsFlag::kFlutterSemanticsFlagIsTextField |
-      FlutterSemanticsFlag::kFlutterSemanticsFlagIsFocused);
+  FlutterSemanticsFlags flags = FlutterSemanticsFlags{0};
+  flags.isTextField = true;
+  flags.isReadOnly = true;
```
Use initializer list syntax here:
```cc
auto flags = FlutterSemanticsFlags{
  .is_text_field,
  .is_readonly,
};
```

cbracken reviewed (changes requested):
A few big questions:
* For the embedder API, how are we going to populate the old `flags` field so that we don't break existing embedders that were built against it, but drop in a new engine shared library?
* This looks like it doesn't modify the Windows/Linux embedders to use flags2, is that intentional?

Other than that, mostly small C++ nits/suggestions!

hannah-hyj committed:
Update engine/src/flutter/shell/platform/embedder/embedder_semantics_update.cc

Co-authored-by: Chris Bracken <chris@bracken.jp>

hannah-hyj committed:
resolve some comments

resolve some comments

Update flutter_platform_node_delegate_unittests.cc

update windows tests

struct_size

resolve some comments

hannah-hyj commented:
>>For the embedder API, how are we going to populate the old flags field so that we don't break existing embedders that were built against it, but drop in a new engine shared library?

We still convert the first 30 flags to an int and populate the old flag. There's a function  `SemanticsFlagsToInt` in `embedder_semantics_update.cc ` for it, but for any new flags, it will only be added to the flags2 struct.



>>This looks like it doesn't modify the Windows/Linux embedders to use flags2, is that intentional? 


I think I just missed them 😅 , where are the Windows/Linux embedders located?

hannah-hyj commented:
@cbracken 

For Linux, I didn't update it but I think we don't own the code anymore.
For Windows,  I updated all usage of `FlutterSemanticsFlag ` under engine/src/flutter/shell/platform/windows/ and it's just some tests

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -258,6 +258,41 @@ typedef enum {
 } FlutterSemanticsFlag;
 
+typedef struct {
+  size_t struct_size;
```
```suggestion
  /// The size of this struct. Must be sizeof(FlutterSemanticsFlags).
  size_t struct_size;
```

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -258,6 +258,41 @@ typedef enum {
 
+typedef struct {
+  size_t struct_size;
+  bool has_checked_state;
```
For all these members, could we copy the comment from `FlutterSemanticsFlag`?

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder_semantics_update.cc:
```diff
@@ -279,6 +321,7 @@ void EmbedderSemanticsUpdate2::AddNode(const SemanticsNode& node) {
       increased_value_attributes.attributes,
       decreased_value_attributes.count,
       decreased_value_attributes.attributes,
+      &flags_.back(),
```
While I believe the current logic is safe because we called [`flags_.reserve(nodes.size())`](https://github.com/flutter/flutter/pull/167738/files#diff-e6b6c0da985d95fc22f4f4faf1286a70a95103cdc3017ec5bbe6521e5aa5fb10R234), I worry this might result in a crash in the future if we ever add more `flags_` than expected - this would cause the vector to resize, and would invalidate all the pointers we've captured.

I see that for nodes and actions we did a [two-phased approach](https://github.com/flutter/flutter/blob/25e8e074cfd1a18b7016cf72ecebf7f68ea90103/engine/src/flutter/shell/platform/embedder/embedder_semantics_update.cc#L194-L213) to be extra safe:

1. We start by adding all the nodes and actions to a vector
2. After all nodes and actions are fully built, we capture pointers to the nodes and actions

Maybe we should be extra safe with `flags_` as well? We could make `flags_` a `std::vector<std::unique_ptr<FlutterSemanticsFlags>>`. This would ensure that the `FlutterSemanticsFlags*` does not get invalidated even if we have an unexpected number of `flags_`? I might be overthinking this though, what do you think @cbracken?

cbracken replied:
Agreed that the current logic is safe, for the reason you state. Agreed that it makes me uncomfortable that we're deep inside another function, nearly 100 lines away from the `reserve` call, which is critical to this approach being safe. If we were just a couple lines away, I'd be much more comfortable with this.

What @loic-sharma is suggesting is probably the safest idea. Have the `ConvertToFlutterSemanticsFlags` do `return std::make_unique<FlutterSemanticsFlags>(...);`.

You'll then grab a raw pointer and move it into the array:
```cc
auto flags = ConvertToFlutterSemanticsFlags(node.flags);
FlutterSemanticsFlags* flags_ptr = flags.get();
flags_.push_back(std::move(flags));
nodes_.push_back(
  // ...
  flags_ptr,
);
```

loic-sharma commented on engine/src/flutter/shell/platform/windows/flutter_windows_view_unittests.cc:
```diff
@@ -162,9 +162,12 @@ TEST(FlutterWindowsViewTest, SubMenuExpandedState) {
-  root.flags = static_cast<FlutterSemanticsFlag>(
-      FlutterSemanticsFlag::kFlutterSemanticsFlagHasExpandedState |
-      FlutterSemanticsFlag::kFlutterSemanticsFlagIsExpanded);
+  auto flags = FlutterSemanticsFlags{
+      .has_expanded_state = true,
+      .is_expanded = true,
+  };
```
My C++ is getting rusty already, but I believe our style is:

```suggestion
  FlutterSemanticsFlags flags = {
      .has_expanded_state = true,
      .is_expanded = true,
  };
```

Same comment for below

loic-sharma replied:
Oh nope never mind, Chris posted above that this is the correct style. Definitely prefer Chris's comments over mine. Closing!

loic-sharma commented on engine/src/flutter/shell/platform/common/accessibility_bridge_unittests.cc:
```diff
@@ -209,9 +209,15 @@ TEST(AccessibilityBridgeTest, CanHandleSelectionChangeCorrectly) {
-      FlutterSemanticsFlag::kFlutterSemanticsFlagIsTextField |
-      FlutterSemanticsFlag::kFlutterSemanticsFlagIsFocused);
+
+  auto flags = FlutterSemanticsFlags{
+
+      .is_text_field = true,
+      .is_read_only = true,
+  };
```
I believe this matches our style guide:

```suggestion
  FlutterSemanticsFlags flags = {
      .is_text_field = true,
      .is_read_only = true,
  };
```

hannah-hyj committed:
Update engine/src/flutter/shell/platform/embedder/embedder.h

Co-authored-by: Loïc Sharma <737941+loic-sharma@users.noreply.github.com>

hannah-hyj committed:
Update engine/src/flutter/shell/platform/common/accessibility_bridge_unittests.cc

Co-authored-by: Loïc Sharma <737941+loic-sharma@users.noreply.github.com>

hannah-hyj committed:
Copy comments

hannah-hyj committed:
add deprecated comments

hannah-hyj committed:
resolve comments &rename the old flag to flags__deprecated__

resolve comments

resolve comments &rename the old flag to flags__deprecated__

cbracken commented on engine/src/flutter/shell/platform/embedder/embedder_semantics_update.cc:
```diff
@@ -238,6 +278,8 @@ void EmbedderSemanticsUpdate2::AddNode(const SemanticsNode& node) {
   auto decreased_value_attributes =
       CreateStringAttributes(node.decreasedValueAttributes);
+  FlutterSemanticsFlags flags = ConvertToFlutterSemanticsFlags(node.flags);
+  flags_.push_back(flags);
```
In theory, I _think_ you can do this to construct in-place in the vector, or at least take advantage of RVO in C++:
```cc
flags_.emplace_back(ConvertToFlutterSemanticsFlags(node.flags));
```

See below, since I have the same concern as @loic-sharma about returning a raw pointer into the flags array. If we go with the `unique_ptr` approach, this may be equally useful.

cbracken commented on engine/src/flutter/shell/platform/embedder/embedder_semantics_update.cc:
```diff
@@ -279,6 +321,7 @@ void EmbedderSemanticsUpdate2::AddNode(const SemanticsNode& node) {
       increased_value_attributes.attributes,
       decreased_value_attributes.count,
       decreased_value_attributes.attributes,
+      &flags_.back(),
```
GitHub won't let me comment in places where there isn't a diff, but I'd be tempted to add a comment to the `SemanticsNodeToInt()` function above -- similar to the one on the `flags` field itself, that states that it's only for backward compatibility and that it only contains a subset of the flags.

cbracken reviewed (commented):
Overall this is looking really good. Just a few last nits, and I think we're good to go.

Are there tests in https://github.com/flutter/flutter/blob/master/engine/src/flutter/shell/platform/embedder/tests/embedder_a11y_unittests.cc that need updating to test the new flags?

hannah-hyj committed:
Update FlutterPlatformNodeDelegateMacTest.mm

hannah-hyj committed:
Update fl_view_accessible.cc

hannah-hyj committed:
fix some tests

Update flutter_platform_node_delegate_unittests.cc

update test

fix tests

hannah-hyj committed:
Update flutter_platform_node_delegate_unittests.cc

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -258,6 +262,94 @@ typedef enum {
+typedef struct {
+  /// The size of this struct. Must be sizeof(FlutterSemanticsFlags).
+  size_t struct_size;
+  /// The semantics node has the quality of either being "checked" or
+  /// "unchecked".
+  bool has_checked_state;
+  /// Whether a semantics node is checked.
+  bool is_checked;
```
Do we want to do something nicer here? We could have a single member for checked state if we introduced an enum with values "none", "true", "false". I don't feel strongly about this at all, feel free to skip.

hannah-hyj replied:
but i think it can be in another PR to update embedder and engine and frameworrk all together.

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -1594,6 +1690,9 @@ typedef struct {
   // Has length `decreased_value_attribute_count`.
   const FlutterStringAttribute** decreased_value_attributes;
+  // The set of semantics flags associated with this node. Prefer to use this
+  // over `flags`.
```
```suggestion
  // over `flags__deprecated__`.
```

loic-sharma reviewed (approved):
Looks good, nice work!

robert-ancell commented on engine/src/flutter/shell/platform/linux/fl_view_accessible.cc:
```diff
@@ -33,7 +33,7 @@ static FlAccessibleNode* create_node(FlViewAccessible* self,
   }
 
-  if (semantics->flags & kFlutterSemanticsFlagIsTextField) {
+  if (semantics->flags__deprecated__ & kFlutterSemanticsFlagIsTextField) {
```
Ack, I'll take a task to migrate to flags2 when this lands.

robert-ancell replied:
See https://github.com/flutter/flutter/pull/170497

hannah-hyj committed:
Update engine/src/flutter/shell/platform/embedder/embedder.h

Co-authored-by: Loïc Sharma <737941+loic-sharma@users.noreply.github.com>

cbracken reviewed (approved):
![LGTM stamp from a Japanese personal seal](https://github.com/user-attachments/assets/61009e3c-6bf6-494d-afb4-d56aac3b61d4)

LGTM -- awesome work! Thanks so much for the backwards-compatibility/deprecation work on this!

hannah-hyj committed:
Merge branch 'master' into semanticsflags-embedder

hannah-hyj committed:
update tests

Update flutter_windows_view_unittests.cc

hannah-hyj committed:
Update accessibility_bridge_windows_unittests.cc

hannah-hyj committed:
fix tests

hannah-hyj committed:
Merge branch 'master' into semanticsflags-embedder

hannah-hyj committed:
Update FlutterPlatformNodeDelegateMacTest.mm

hannah-hyj commented:
Reason for revert: want to make more changes about tristate flags.
